pkg/services/internal/servicequeue/queues.go

```diff
@@ -0,0 +1,193 @@
+	logger = logging.Child(logger, "services."+methodName)
+
+	return Queue2[A, B, R]{
+		logger:     logger,
+		methodName: methodName,
+		services:   services,
+	}
+}
+
+// OneByOne processes services one by one until a successful result is obtained.
+// The context and arguments are passed to each service.
+// Returns the first successful result or an error if all services fail.
+func (q *Queue2[A, B, R]) OneByOne(ctx context.Context, a A, b B) (R, error) {
+	return processOneByOne(q.logger, q.services, func(s *Service2[A, B, R]) (R, error) {
+		return s.service(ctx, a, b)
+	})
+}
+
+// Queue3 is a structure that holds a collection of services and abstracts away the details of calling them and error handling.
+// Services are functions accepting a context and three arguments and returning a result or an error.
+// For different number of arguments see Queue, Queue1, and Queue2.
+type Queue3[A, B, C, R any] struct {
+	logger     *slog.Logger
+	methodName string
+	services   []*Service3[A, B, C, R]
+}
+
+func NewQueue3[A, B, C, R any](logger *slog.Logger, methodName string, services ...*Service3[A, B, C, R]) Queue3[A, B, C, R] {
+	logger = logging.Child(logger, "services."+methodName)
+
+	return Queue3[A, B, C, R]{
+		logger:     logger,
+		methodName: methodName,
+		services:   services,
+	}
+}
+
+// OneByOne processes services one by one until a successful result is obtained.
+// The context and arguments are passed to each service.
+// Returns the first successful result or an error if all services fail.
+func (q *Queue3[A, B, C, R]) OneByOne(ctx context.Context, a A, b B, c C) (R, error) {
+	return processOneByOne(q.logger, q.services, func(s *Service3[A, B, C, R]) (R, error) {
+		return s.service(ctx, a, b, c)
+	})
+}
+
+type serv interface {
+	Name() string
+}
+
+func processOneByOne[S serv, R any](logger *slog.Logger, services []S, callService func(S) (R, error)) (R, error) {
+	if len(services) == 0 {
+		return to.ZeroValue[R](), errors.New("no services registered")
+	}
+
+	results := seq.MapOrErr(seq.FromSlice(services), func(s S) (collection.Tuple2[string, R], error) {
+		res, err := callService(s)
+		if err != nil {
+			err = fmt.Errorf("service %s responds with error: %w", s.Name(), err)
+		}
+		return collection.NewTuple2(s.Name(), res), err
+	})
+
+	results = takeUntilHaveResult(results)
+
+	results = seq2.Each(results, func(serviceResult collection.Tuple2[string, R], err error) {
+		if err != nil {
+			logger.Warn("error when calling service",
+				slog.String("service.name", serviceResult.A),
+				logging.Error(err),
+			)
+		}
+	})
+
+	result := seq2.Reduce(results,
+		func(res collection.Tuple2[R, error], serviceResult collection.Tuple2[string, R], err error) collection.Tuple2[R, error] {
+			ret := collection.NewTuple2[R, error](serviceResult.B, errors.Join(res.B, err))
+			return ret
+		},
+		collection.Tuple2[R, error]{},
+	)
```

Comment:
It looks quite complex - and uses external methods that a reader needs to know what they do.

Consider more straight-forward way.
```suggestion
	result := collection.Tuple2[R, error]{} // tuple could be also changed to actual struct
	for serviceResult, err := range takeUntilHaveResult(results) {
		if err != nil {
			logger.Warn("error when calling service",
				slog.String("service.name", serviceResult.A),
				logging.Error(err),
			)
		}
		result = collection.NewTuple2[R, error](serviceResult.B, errors.Join(result.B, err))
	}
```

Reply:
It's a matter of taste, and in my opinion your suggestion aren't that much clearer or something.

> and uses external methods that a reader needs to know

The reduce method is popular in many languages as a part of filter/map/reduce functional operations, 
and for me, introducing iter.Seq in go, opens the doors to use this pattern, therefore I don't see any reason to not use this pattern just because a reader need to know the method Reduce, or something.

Reply:
But I play with the last part and error handling and can have something simpler.